precompile/assetsbridge/bridge_out.go

```diff
@@ -0,0 +1,443 @@
+	return 0, false
+}
+
+func (m *BridgeOutMethod) Payable() bool {
+	return false
+}
+
+func (m *BridgeOutMethod) Run(
+	context *precompile.RunContext,
+	rawInputs precompile.MethodInputs,
+) (precompile.MethodOutputs, error) {
+	// extract inputs
+	inputs, err := m.extractInputs(rawInputs)
+	if err != nil {
+		return precompile.MethodOutputs{false}, err
+	}
+
+	// run validation
+	if err := m.validate(context, inputs); err != nil {
+		return precompile.MethodOutputs{false}, err
+	}
+
+	return m.execute(context, inputs)
+}
+
+// execute will execute the burn of the token then send
+// AssetsUnlocked events to the bridgeKeeper.
+func (m *BridgeOutMethod) execute(
+	context *precompile.RunContext,
+	inputs *bridgeOutInputs,
+) (precompile.MethodOutputs, error) {
+	var (
+		err            error
+		assetsUnlocked *bridgetypes.AssetsUnlockedEvent
+		isBTC          = bytes.Equal(
+			common.HexToAddress(evmtypes.BTCTokenPrecompileAddress).Bytes(),
+			inputs.Token.Bytes(),
+		)
+	)
+
+	switch inputs.Chain {
+	case TargetChainEthereum:
+		if isBTC {
+			assetsUnlocked, err = m.executeBitcoin(context, inputs)
+		} else {
+			assetsUnlocked, err = m.executeEthereum(context, inputs)
+		}
+	case TargetChainBitcoin:
+		assetsUnlocked, err = m.executeBitcoin(context, inputs)
+	}
+
+	if assetsUnlocked != nil {
+		err := context.EventEmitter().Emit(
+			NewAssetsUnlockedEvent(
+				context.MsgSender(),
+				common.HexToAddress(assetsUnlocked.Token),
+				assetsUnlocked.Recipient,
+				uint8(assetsUnlocked.Chain), //nolint:gosec // G115: Safe conversion, Chain is validated elsewhere
+				assetsUnlocked.UnlockSequence.BigInt(),
+				assetsUnlocked.Amount.BigInt(),
+			),
+		)
+		if err != nil {
+			return nil, fmt.Errorf("failed to emit AssetsUnlocked event: [%w]", err)
+		}
+
+	}
+
+	return precompile.MethodOutputs{err == nil}, err
+}
+
+func (m *BridgeOutMethod) executeEthereum(
+	context *precompile.RunContext,
+	inputs *bridgeOutInputs,
+) (*bridgetypes.AssetsUnlockedEvent, error) {
+	var (
+		sdkCtx          = context.SdkCtx()
+		bridgeAddrBytes = common.HexToAddress(
+			evmtypes.AssetsBridgePrecompileAddress,
+		).Bytes()
+		spenderAddr = context.MsgSender()
+	)
+
+	sdkAmount, err := precompile.TypesConverter.BigInt.ToSDK(inputs.Amount)
+	if err != nil {
+		return nil, fmt.Errorf("unable to convert amount to sdk type: %v", err)
+	}
+
+	call, err := evmtypes.NewERC20BurnFromCall(
+		bridgeAddrBytes,
+		inputs.Token.Bytes(),
+		spenderAddr.Bytes(),
+		inputs.Amount,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create ERC20 burnFrom call: %w", err)
+	}
+
+	_, err = m.evmKeeper.ExecuteContractCall(sdkCtx, call)
+	if err != nil {
+		return nil, fmt.Errorf("failed to execute ERC20 burnFrom call: %w", err)
+	}
+
+	assetsUnlocked, err := m.bridgeKeeper.AssetsUnlocked(
+		sdkCtx,
+		inputs.Token.Bytes(),
+		sdkAmount,
+		uint8(inputs.Chain),
+		inputs.Recipient,
+	)
+	if err != nil {
+		// TODO(JEREMY): error happened here, funds might have been
```

Comment:
What do we do about this TODO?
Are we going to address it in this or the following TODO?

Reply:
that's something I'd say is open for discussion still. I would panic to be fair. if you can't write it down to the state, then something v wrong would have happened, and I'd panic to investigate. but I guess it'd be good for you guys to feedback on what you think.

Reply:
removed the comment, and keeping as is then. 5451bf62